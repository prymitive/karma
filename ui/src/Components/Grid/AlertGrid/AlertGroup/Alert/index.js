import React from "react";
import PropTypes from "prop-types";

import { useObserver } from "mobx-react-lite";

import { APIAlert, APIGroup } from "Models/API";
import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { BorderClassMap } from "Common/Colors";
import { StaticLabels } from "Common/Query";
import { FilteringLabel } from "Components/Labels/FilteringLabel";
import { InhibitedByModal } from "Components/InhibitedByModal";
import { RenderNonLinkAnnotation, RenderLinkAnnotation } from "../Annotation";
import { AlertMenu } from "./AlertMenu";
import { RenderSilence } from "../Silences";

const Alert = ({
  group,
  alert,
  showAlertmanagers,
  showReceiver,
  afterUpdate,
  alertStore,
  silenceFormStore,
  setIsMenuOpen,
}) => {
  const classNames = [
    "components-grid-alertgrid-alertgroup-alert",
    "list-group-item bg-transparent",
    "pl-1 pr-0 py-0",
    "my-1",
    "rounded-0",
    "border-left-1 border-right-0 border-top-0 border-bottom-0",
    BorderClassMap[alert.state] || "border-default",
  ];

  const silences = {};
  let clusters = [];
  let inhibitedBy = [];
  for (const am of alert.alertmanager) {
    if (!clusters.includes(am.cluster)) {
      clusters.push(am.cluster);
    }
    for (const fingerprint of am.inhibitedBy) {
      if (!inhibitedBy.includes(fingerprint)) {
        inhibitedBy.push(fingerprint);
      }
    }
    if (!silences[am.cluster]) {
      silences[am.cluster] = {
        alertmanager: am,
        silences: [
          ...new Set(
            am.silencedBy.filter(
              (silenceID) =>
                !(
                  group.shared.silences[am.cluster] &&
                  group.shared.silences[am.cluster].includes(silenceID)
                )
            )
          ),
        ],
      };
    }
  }

  return useObserver(() => (
    <li className={classNames.join(" ")}>
      <div>
        {alert.annotations
          .filter((a) => a.isLink === false)
          .map((a) => (
            <RenderNonLinkAnnotation
              key={a.name}
              name={a.name}
              value={a.value}
              visible={a.visible}
              afterUpdate={afterUpdate}
            />
          ))}
      </div>
      <AlertMenu
        group={group}
        alert={alert}
        alertStore={alertStore}
        silenceFormStore={silenceFormStore}
        setIsMenuOpen={setIsMenuOpen}
      />
      {inhibitedBy.length > 0 ? (
        <InhibitedByModal alertStore={alertStore} fingerprints={inhibitedBy} />
      ) : null}
      {Object.entries(alert.labels).map(([name, value]) => (
        <FilteringLabel
          key={name}
          name={name}
          value={value}
          alertStore={alertStore}
        />
      ))}
      {showAlertmanagers
        ? clusters.map((cluster) => (
            <FilteringLabel
              key={cluster}
              name={StaticLabels.AlertmanagerCluster}
              value={cluster}
              alertStore={alertStore}
            />
          ))
        : null}
      {showReceiver ? (
        <FilteringLabel
          name={StaticLabels.Receiver}
          value={alert.receiver}
          alertStore={alertStore}
        />
      ) : null}
      {alert.annotations
        .filter((a) => a.isLink === true)
        .map((a) => (
          <RenderLinkAnnotation key={a.name} name={a.name} value={a.value} />
        ))}
      {Object.entries(silences).map(([cluster, clusterSilences]) =>
        clusterSilences.silences.map((silenceID) =>
          RenderSilence(
            alertStore,
            silenceFormStore,
            afterUpdate,
            cluster,
            silenceID
          )
        )
      )}
    </li>
  ));
};
Alert.propTypes = {
  group: APIGroup.isRequired,
  alert: APIAlert.isRequired,
  showAlertmanagers: PropTypes.bool.isRequired,
  showReceiver: PropTypes.bool.isRequired,
  afterUpdate: PropTypes.func.isRequired,
  alertStore: PropTypes.instanceOf(AlertStore).isRequired,
  silenceFormStore: PropTypes.instanceOf(SilenceFormStore).isRequired,
  setIsMenuOpen: PropTypes.func.isRequired,
};

export { Alert };
