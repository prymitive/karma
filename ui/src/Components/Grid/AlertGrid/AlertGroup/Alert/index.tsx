import { FC } from "react";

import { observer } from "mobx-react-lite";

import {
  APIAlertT,
  APIAlertGroupT,
  APIAlertmanagerStateT,
} from "Models/APITypes";
import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { BorderClassMap } from "Common/Colors";
import { StaticLabels } from "Common/Query";
import FilteringLabel from "Components/Labels/FilteringLabel";
import { InhibitedByModal } from "Components/InhibitedByModal";
import { RenderNonLinkAnnotation, RenderLinkAnnotation } from "../Annotation";
import { AlertMenu } from "./AlertMenu";
import { RenderSilence } from "../Silences";

const Alert: FC<{
  group: APIAlertGroupT;
  alert: APIAlertT;
  showAlertmanagers: boolean;
  showReceiver: boolean;
  showOnlyExpandedAnnotations: boolean;
  afterUpdate: () => void;
  alertStore: AlertStore;
  silenceFormStore: SilenceFormStore;
  setIsMenuOpen: (isOpen: boolean) => void;
}> = ({
  group,
  alert,
  showAlertmanagers,
  showReceiver,
  showOnlyExpandedAnnotations,
  afterUpdate,
  alertStore,
  silenceFormStore,
  setIsMenuOpen,
}) => {
  const classNames = [
    "components-grid-alertgrid-alertgroup-alert",
    "list-group-item bg-transparent",
    "ps-1 pe-0 py-0",
    "my-1",
    "rounded-0",
    "border-start-1 border-end-0 border-top-0 border-bottom-0",
    BorderClassMap[alert.state] || "border-default",
  ];

  const silences: {
    [cluster: string]: {
      alertmanager: APIAlertmanagerStateT;
      silences: string[];
    };
  } = {};
  const clusters: string[] = [];
  const inhibitedBy: string[] = [];
  for (const am of alert.alertmanager) {
    if (!clusters.includes(am.cluster)) {
      clusters.push(am.cluster);
    }
    for (const fingerprint of am.inhibitedBy) {
      if (!inhibitedBy.includes(fingerprint)) {
        inhibitedBy.push(fingerprint);
      }
    }
    if (!silences[am.cluster]) {
      silences[am.cluster] = {
        alertmanager: am,
        silences: Array.from(
          new Set(
            am.silencedBy.filter(
              (silenceID) =>
                !(
                  group.shared.silences[am.cluster] &&
                  group.shared.silences[am.cluster].includes(silenceID)
                )
            )
          )
        ),
      };
    }
  }

  return (
    <li className={classNames.join(" ")}>
      <div>
        {alert.annotations
          .filter((a) => a.isLink === false)
          .filter((a) => a.visible === true || !showOnlyExpandedAnnotations)
          .map((a) => (
            <RenderNonLinkAnnotation
              key={a.name}
              name={a.name}
              value={a.value}
              visible={a.visible}
              allowHTML={alertStore.settings.values.annotationsEnableHTML}
              afterUpdate={afterUpdate}
            />
          ))}
      </div>
      <AlertMenu
        group={group}
        alert={alert}
        alertStore={alertStore}
        silenceFormStore={silenceFormStore}
        setIsMenuOpen={setIsMenuOpen}
      />
      {inhibitedBy.length > 0 ? (
        <InhibitedByModal alertStore={alertStore} fingerprints={inhibitedBy} />
      ) : null}
      {Object.entries(alert.labels).map(([name, value]) => (
        <FilteringLabel
          key={name}
          name={name}
          value={value}
          alertStore={alertStore}
        />
      ))}
      {showAlertmanagers
        ? clusters.map((cluster) => (
            <FilteringLabel
              key={cluster}
              name={StaticLabels.AlertmanagerCluster}
              value={cluster}
              alertStore={alertStore}
            />
          ))
        : null}
      {showReceiver ? (
        <FilteringLabel
          name={StaticLabels.Receiver}
          value={alert.receiver}
          alertStore={alertStore}
        />
      ) : null}
      {alert.annotations
        .filter((a) => a.isLink === true)
        .filter((a) => a.isAction === false)
        .map((a) => (
          <RenderLinkAnnotation key={a.name} name={a.name} value={a.value} />
        ))}
      {Object.entries(silences).map(([cluster, clusterSilences]) =>
        clusterSilences.silences.map((silenceID) => (
          <RenderSilence
            key={silenceID}
            alertStore={alertStore}
            silenceFormStore={silenceFormStore}
            afterUpdate={afterUpdate}
            cluster={cluster}
            silenceID={silenceID}
          />
        ))
      )}
    </li>
  );
};

export default observer(Alert);
