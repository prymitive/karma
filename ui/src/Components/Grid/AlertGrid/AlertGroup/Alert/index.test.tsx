import { act } from "react-dom/test-utils";

import { mount } from "enzyme";

import { advanceTo, clear } from "jest-date-mock";

import toDiffableHtml from "diffable-html";

import {
  MockAlert,
  MockAnnotation,
  MockAlertGroup,
  MockSilence,
} from "__fixtures__/Alerts";
import { MockThemeContext } from "__fixtures__/Theme";
import { APIAlertGroupT, APIAlertT } from "Models/APITypes";
import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { BorderClassMap } from "Common/Colors";
import { ThemeContext } from "Components/Theme";
import Alert from ".";

let alertStore: AlertStore;
let silenceFormStore: SilenceFormStore;

beforeEach(() => {
  advanceTo(new Date(Date.UTC(2018, 7, 15, 20, 40, 0)));
  alertStore = new AlertStore([]);
  silenceFormStore = new SilenceFormStore();
});

afterEach(() => {
  // reset Date() to current time
  clear();
  jest.restoreAllMocks();
});

const MockAfterUpdate = jest.fn();
const MockSetIsMenuOpen = jest.fn();

const MockedAlert = () => {
  return MockAlert(
    [
      MockAnnotation("help", "some long text", true, false, false),
      MockAnnotation("hidden", "some hidden text", false, false, false),
      MockAnnotation("link", "http://localhost", true, true, false),
    ],
    { job: "node_exporter", cluster: "dev" },
    "active"
  );
};

const MountedAlert = (
  alert: APIAlertT,
  group: APIAlertGroupT,
  showAlertmanagers: boolean,
  showReceiver: boolean,
  showOnlyExpandedAnnotations: boolean
) => {
  return mount(
    <Alert
      alert={alert}
      group={group}
      showAlertmanagers={showAlertmanagers}
      showReceiver={showReceiver}
      showOnlyExpandedAnnotations={showOnlyExpandedAnnotations}
      afterUpdate={MockAfterUpdate}
      alertStore={alertStore}
      silenceFormStore={silenceFormStore}
      setIsMenuOpen={MockSetIsMenuOpen}
    />,
    {
      wrappingComponent: ThemeContext.Provider,
      wrappingComponentProps: { value: MockThemeContext },
    }
  );
};

describe("<Alert />", () => {
  it("matches snapshot with showAlertmanagers=false showReceiver=false", () => {
    const alert = MockedAlert();
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("matches snapshot when inhibited", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].inhibitedBy = ["123456"];
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("renders inhibition icon when inhibited", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].inhibitedBy = ["123456"];
    alert.alertmanager.push({
      fingerprint: "abc",
      name: "ha2",
      cluster: "HA",
      state: "active",
      startsAt: "2018-08-14T17:36:40.017867056Z",
      source: "http://localhost/graph",
      silencedBy: [],
      inhibitedBy: ["123456"],
    });
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    expect(tree.find(".fa-volume-mute")).toHaveLength(1);
  });

  it("inhibition icon passes only unique fingerprints", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].inhibitedBy = ["123456"];
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    expect(tree.find(".fa-volume-mute")).toHaveLength(1);
  });

  it("renders @cluster label with showAlertmanagers=true", () => {
    const alert = MockedAlert();
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, true, false, false);
    const label = tree
      .find("FilteringLabel")
      .filterWhere((elem) => elem.props().name === "@cluster");
    expect(label.text()).toBe("@cluster: default");
  });

  it("only renders one @cluster label per alertmanager cluster", () => {
    const alert = MockedAlert();
    alert.alertmanager.push({
      fingerprint: "123",
      name: "ha1",
      cluster: "HA",
      state: "active",
      startsAt: "2018-08-14T17:36:40.017867056Z",
      source: "http://localhost/graph",
      silencedBy: [],
      inhibitedBy: [],
    });
    alert.alertmanager.push({
      fingerprint: "123",
      name: "ha2",
      cluster: "HA",
      state: "active",
      startsAt: "2018-08-14T17:36:40.017867056Z",
      source: "http://localhost/graph",
      silencedBy: [],
      inhibitedBy: [],
    });
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, true, false, false);
    const labels = tree
      .find("FilteringLabel")
      .filterWhere((elem) => elem.props().name === "@cluster");
    expect(labels).toHaveLength(2);
    expect(labels.at(0).text()).toBe("@cluster: default");
    expect(labels.at(1).text()).toBe("@cluster: HA");
  });

  it("renders @receiver label with showReceiver=true", () => {
    const alert = MockedAlert();
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, true, false);
    const label = tree
      .find("FilteringLabel")
      .filterWhere((elem) => elem.props().name === "@receiver");
    expect(label.text()).toBe("@receiver: by-name");
  });

  it("renders a silence if alert is silenced", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].silencedBy = ["silence123456789"];
    alertStore.data.setSilences({
      default: {
        silence123456789: MockSilence(),
      },
    });
    const group = MockAlertGroup({}, [alert], [], {}, { default: [] });
    const tree = MountedAlert(alert, group, false, false, false);
    const silence = tree.find("ManagedSilence");
    expect(silence).toHaveLength(1);
    expect(silence.html()).toMatch(/Mocked Silence/);
  });

  it("renders a fallback silence if the silence is not found in alertStore", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].silencedBy = ["silence123456789"];
    alertStore.data.setSilences({
      default: {
        "123": MockSilence(),
      },
    });
    const group = MockAlertGroup({}, [alert], [], {}, { default: [] });
    const tree = MountedAlert(alert, group, false, false, false);
    const silence = tree.find("FallbackSilenceDesciption");
    expect(silence).toHaveLength(1);
    expect(silence.html()).not.toMatch(/Mocked Silence/);
  });

  it("renders a fallback silence if the cluster is not found in alertStore", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].silencedBy = ["silence123456789"];
    alertStore.data.setSilences({
      foo: {
        "123": MockSilence(),
      },
    });
    const group = MockAlertGroup({}, [alert], [], {}, { default: [] });
    const tree = MountedAlert(alert, group, false, false, false);
    const silence = tree.find("FallbackSilenceDesciption");
    expect(silence).toHaveLength(1);
    expect(silence.html()).not.toMatch(/Mocked Silence/);
  });

  it("renders only one silence for HA cluster", () => {
    const alert = MockedAlert();
    alert.alertmanager = [
      {
        fingerprint: "123",
        name: "am1",
        cluster: "ha",
        state: "suppressed",
        startsAt: "2018-08-14T17:36:40.017867056Z",
        source: "localhost/am1",
        silencedBy: ["silence123456789"],
        inhibitedBy: [],
      },
      {
        fingerprint: "123",
        name: "am2",
        cluster: "ha",
        state: "suppressed",
        startsAt: "2018-08-14T17:36:40.017867056Z",
        source: "localhost/am2",
        silencedBy: ["silence123456789"],
        inhibitedBy: [],
      },
    ];
    alertStore.data.setSilences({
      ha: {
        silence123456789: MockSilence(),
      },
    });
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    const silence = tree.find("ManagedSilence");
    expect(silence).toHaveLength(1);
    expect(silence.html()).toMatch(/Mocked Silence/);
  });

  it("doesn't render shared silences", () => {
    const alert = MockedAlert();
    alert.alertmanager[0].silencedBy = ["silence123456789"];
    const group = MockAlertGroup(
      {},
      [alert],
      [],
      {},
      { default: ["silence123456789"] }
    );
    const tree = MountedAlert(alert, group, false, false, false);
    const silence = tree.find("ManagedSilence");
    expect(silence).toHaveLength(0);
  });

  it("renders collapsed annotations when showOnlyExpandedAnnotations=false", () => {
    const alert = MockedAlert();
    alert.annotations = [
      {
        name: "visible",
        value: "1",
        visible: true,
        isLink: false,
        isAction: false,
      },
      {
        name: "invisible",
        value: "2",
        visible: false,
        isLink: false,
        isAction: false,
      },
    ];
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    const annotations = tree.find("div.components-grid-annotation");
    expect(annotations).toHaveLength(2);
  });

  it("doesn't render collapsed annotations when showOnlyExpandedAnnotations=true", () => {
    const alert = MockedAlert();
    alert.annotations = [
      {
        name: "visible",
        value: "1",
        visible: true,
        isLink: false,
        isAction: false,
      },
      {
        name: "invisible",
        value: "2",
        visible: false,
        isLink: false,
        isAction: false,
      },
    ];
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, true);
    const annotations = tree.find("div.components-grid-annotation");
    expect(annotations).toHaveLength(1);
  });

  it("uses BorderClassMap.active when @state=active", () => {
    const alert = MockedAlert();
    alert.state = "active";
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    expect(
      tree
        .find(".components-grid-alertgrid-alertgroup-alert")
        .hasClass(BorderClassMap.active)
    ).toBe(true);
  });

  it("uses BorderClassMap.suppressed when @state=suppressed", () => {
    const alert = MockedAlert();
    alert.state = "suppressed";
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    expect(
      tree
        .find(".components-grid-alertgrid-alertgroup-alert")
        .hasClass(BorderClassMap.suppressed)
    ).toBe(true);
  });

  it("uses BorderClassMap.unprocessed when @state=unprocessed", () => {
    const alert = MockedAlert();
    alert.state = "unprocessed";
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    expect(
      tree
        .find(".components-grid-alertgrid-alertgroup-alert")
        .hasClass(BorderClassMap.unprocessed)
    ).toBe(true);
  });

  it("uses 'border-default' with unknown @state", () => {
    jest.spyOn(console, "error").mockImplementation(() => {});

    const alert = MockedAlert();
    (alert.state as string) = "foobar";
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    expect(
      tree
        .find(".components-grid-alertgrid-alertgroup-alert")
        .hasClass("border-default")
    ).toBe(true);
  });

  it("alert timestamp is updated every minute", () => {
    jest.useFakeTimers();

    advanceTo(new Date(Date.UTC(2018, 7, 14, 17, 36, 41)));

    const alert = MockedAlert();
    const group = MockAlertGroup({}, [alert], [], {}, {});
    const tree = MountedAlert(alert, group, false, false, false);
    expect(
      tree
        .find("span.components-label.badge.bg-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("just now");

    advanceTo(new Date(Date.UTC(2018, 7, 14, 17, 36, 42)));
    act(() => {
      jest.advanceTimersByTime(31 * 1000);
    });
    expect(
      tree
        .find("span.components-label.badge.bg-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("a few seconds ago");

    advanceTo(new Date(Date.UTC(2018, 7, 14, 17, 37, 41)));
    act(() => {
      jest.advanceTimersByTime(31 * 1000);
    });
    expect(
      tree
        .find("span.components-label.badge.bg-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("1 minute ago");

    advanceTo(new Date(Date.UTC(2018, 7, 14, 18, 36, 41)));
    act(() => {
      jest.advanceTimersByTime(31 * 1000);
    });
    expect(
      tree
        .find("span.components-label.badge.bg-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("1 hour ago");

    advanceTo(new Date(Date.UTC(2018, 7, 14, 19, 36, 41)));
    act(() => {
      jest.advanceTimersByTime(31 * 1000);
    });
    expect(
      tree
        .find("span.components-label.badge.bg-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("2 hours ago");

    advanceTo(new Date(Date.UTC(2018, 7, 16, 19, 36, 41)));
    act(() => {
      jest.advanceTimersByTime(31 * 1000);
    });
    expect(
      tree
        .find("span.components-label.badge.bg-secondary.cursor-pointer")
        .at(0)
        .text()
    ).toBe("2 days ago");
  });
});
