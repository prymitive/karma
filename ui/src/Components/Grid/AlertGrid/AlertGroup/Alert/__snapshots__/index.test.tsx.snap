// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> matches snapshot when inhibited 1`] = `
"
<li class="components-grid-alertgrid-alertgroup-alert list-group-item bg-transparent ps-1 pe-0 py-0 my-1 rounded-0 border-start-1 border-end-0 border-top-0 border-bottom-0 border-danger">
  <div>
    <div class=" tooltip-trigger">
      <div class="mb-1 p-1 bg-light d-inline-block rounded components-grid-annotation text-break mw-100">
        <span class="cursor-pointer">
          <svg
            aria-hidden="true"
            focusable="false"
            data-prefix="fas"
            data-icon="angle-right"
            class="svg-inline--fa fa-angle-right me-1 text-muted"
            role="img"
            xmlns="http://www.w3.org/2000/svg"
            viewbox="0 0 320 512"
          >
            <path
              fill="currentColor"
              d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"
            >
            </path>
          </svg>
          <span class="text-muted">
            help:
          </span>
        </span>
        <span class="Linkify">
          <span>
            some long text
          </span>
        </span>
      </div>
    </div>
    <div class=" tooltip-trigger">
      <div class="mb-1 p-1 bg-light d-inline-block rounded components-grid-annotation text-break mw-100 cursor-pointer">
        <svg
          aria-hidden="true"
          focusable="false"
          data-prefix="fas"
          data-icon="angle-left"
          class="svg-inline--fa fa-angle-left me-1"
          role="img"
          xmlns="http://www.w3.org/2000/svg"
          viewbox="0 0 320 512"
        >
          <path
            fill="currentColor"
            d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
          >
          </path>
        </svg>
        hidden
      </div>
    </div>
  </div>
  <span>
    <span
      class="components-label components-label-with-hover px-1 me-1 badge bg-secondary cursor-pointer"
      data-toggle="dropdown"
    >
      <svg
        aria-hidden="true"
        focusable="false"
        data-prefix="fas"
        data-icon="caret-down"
        class="svg-inline--fa fa-caret-down pe-1"
        role="img"
        xmlns="http://www.w3.org/2000/svg"
        viewbox="0 0 320 512"
        style="width: 0.8rem;"
      >
        <path
          fill="currentColor"
          d="M137.4 374.6c12.5 12.5 32.8 12.5 45.3 0l128-128c9.2-9.2 11.9-22.9 6.9-34.9s-16.6-19.8-29.6-19.8L32 192c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9l128 128z"
        >
        </path>
      </svg>
      1 day ago
    </span>
  </span>
  <div class=" tooltip-trigger">
    <span class="badge bg-light components-label components-label-with-hover cursor-pointer">
      <svg
        aria-hidden="true"
        focusable="false"
        data-prefix="fas"
        data-icon="volume-xmark"
        class="svg-inline--fa fa-volume-xmark text-success"
        role="img"
        xmlns="http://www.w3.org/2000/svg"
        viewbox="0 0 576 512"
      >
        <path
          fill="currentColor"
          d="M301.1 34.8C312.6 40 320 51.4 320 64V448c0 12.6-7.4 24-18.9 29.2s-25 3.1-34.4-5.3L131.8 352H64c-35.3 0-64-28.7-64-64V224c0-35.3 28.7-64 64-64h67.8L266.7 40.1c9.4-8.4 22.9-10.4 34.4-5.3zM425 167l55 55 55-55c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-55 55 55 55c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-55-55-55 55c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l55-55-55-55c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0z"
        >
        </path>
      </svg>
    </span>
  </div>
  <div class=" tooltip-trigger">
    <span class="components-label badge bg-default components-label-dark components-label-with-hover components-label-without-select">
      <span class="components-label-name">
        job:
      </span>
      <span class="components-label-value">
        node_exporter
      </span>
    </span>
  </div>
  <div class=" tooltip-trigger">
    <span class="components-label badge bg-default components-label-dark components-label-with-hover components-label-without-select">
      <span class="components-label-name">
        cluster:
      </span>
      <span class="components-label-value">
        dev
      </span>
    </span>
  </div>
  <a
    href="http://localhost"
    target="_blank"
    rel="noopener noreferrer"
    class="components-label components-label-with-hover badge border components-grid-annotation-link"
  >
    <svg
      aria-hidden="true"
      focusable="false"
      data-prefix="fas"
      data-icon="up-right-from-square"
      class="svg-inline--fa fa-up-right-from-square "
      role="img"
      xmlns="http://www.w3.org/2000/svg"
      viewbox="0 0 512 512"
    >
      <path
        fill="currentColor"
        d="M352 0c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9L370.7 96 201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L416 141.3l41.4 41.4c9.2 9.2 22.9 11.9 34.9 6.9s19.8-16.6 19.8-29.6V32c0-17.7-14.3-32-32-32H352zM80 32C35.8 32 0 67.8 0 112V432c0 44.2 35.8 80 80 80H400c44.2 0 80-35.8 80-80V320c0-17.7-14.3-32-32-32s-32 14.3-32 32V432c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V112c0-8.8 7.2-16 16-16H192c17.7 0 32-14.3 32-32s-14.3-32-32-32H80z"
      >
      </path>
    </svg>
    link
  </a>
</li>
"
`;

exports[`<Alert /> matches snapshot with showAlertmanagers=false showReceiver=false 1`] = `
"
<li class="components-grid-alertgrid-alertgroup-alert list-group-item bg-transparent ps-1 pe-0 py-0 my-1 rounded-0 border-start-1 border-end-0 border-top-0 border-bottom-0 border-danger">
  <div>
    <div class=" tooltip-trigger">
      <div class="mb-1 p-1 bg-light d-inline-block rounded components-grid-annotation text-break mw-100">
        <span class="cursor-pointer">
          <svg
            aria-hidden="true"
            focusable="false"
            data-prefix="fas"
            data-icon="angle-right"
            class="svg-inline--fa fa-angle-right me-1 text-muted"
            role="img"
            xmlns="http://www.w3.org/2000/svg"
            viewbox="0 0 320 512"
          >
            <path
              fill="currentColor"
              d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"
            >
            </path>
          </svg>
          <span class="text-muted">
            help:
          </span>
        </span>
        <span class="Linkify">
          <span>
            some long text
          </span>
        </span>
      </div>
    </div>
    <div class=" tooltip-trigger">
      <div class="mb-1 p-1 bg-light d-inline-block rounded components-grid-annotation text-break mw-100 cursor-pointer">
        <svg
          aria-hidden="true"
          focusable="false"
          data-prefix="fas"
          data-icon="angle-left"
          class="svg-inline--fa fa-angle-left me-1"
          role="img"
          xmlns="http://www.w3.org/2000/svg"
          viewbox="0 0 320 512"
        >
          <path
            fill="currentColor"
            d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
          >
          </path>
        </svg>
        hidden
      </div>
    </div>
  </div>
  <span>
    <span
      class="components-label components-label-with-hover px-1 me-1 badge bg-secondary cursor-pointer"
      data-toggle="dropdown"
    >
      <svg
        aria-hidden="true"
        focusable="false"
        data-prefix="fas"
        data-icon="caret-down"
        class="svg-inline--fa fa-caret-down pe-1"
        role="img"
        xmlns="http://www.w3.org/2000/svg"
        viewbox="0 0 320 512"
        style="width: 0.8rem;"
      >
        <path
          fill="currentColor"
          d="M137.4 374.6c12.5 12.5 32.8 12.5 45.3 0l128-128c9.2-9.2 11.9-22.9 6.9-34.9s-16.6-19.8-29.6-19.8L32 192c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9l128 128z"
        >
        </path>
      </svg>
      1 day ago
    </span>
  </span>
  <div class=" tooltip-trigger">
    <span class="components-label badge bg-default components-label-dark components-label-with-hover components-label-without-select">
      <span class="components-label-name">
        job:
      </span>
      <span class="components-label-value">
        node_exporter
      </span>
    </span>
  </div>
  <div class=" tooltip-trigger">
    <span class="components-label badge bg-default components-label-dark components-label-with-hover components-label-without-select">
      <span class="components-label-name">
        cluster:
      </span>
      <span class="components-label-value">
        dev
      </span>
    </span>
  </div>
  <a
    href="http://localhost"
    target="_blank"
    rel="noopener noreferrer"
    class="components-label components-label-with-hover badge border components-grid-annotation-link"
  >
    <svg
      aria-hidden="true"
      focusable="false"
      data-prefix="fas"
      data-icon="up-right-from-square"
      class="svg-inline--fa fa-up-right-from-square "
      role="img"
      xmlns="http://www.w3.org/2000/svg"
      viewbox="0 0 512 512"
    >
      <path
        fill="currentColor"
        d="M352 0c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9L370.7 96 201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L416 141.3l41.4 41.4c9.2 9.2 22.9 11.9 34.9 6.9s19.8-16.6 19.8-29.6V32c0-17.7-14.3-32-32-32H352zM80 32C35.8 32 0 67.8 0 112V432c0 44.2 35.8 80 80 80H400c44.2 0 80-35.8 80-80V320c0-17.7-14.3-32-32-32s-32 14.3-32 32V432c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V112c0-8.8 7.2-16 16-16H192c17.7 0 32-14.3 32-32s-14.3-32-32-32H80z"
      >
      </path>
    </svg>
    link
  </a>
</li>
"
`;
